examples/3d/elixir_euler_free_stream_p4est.jl

```diff
@@ -0,0 +1,105 @@
+
+using OrdinaryDiffEq
+using Trixi
+
+###############################################################################
+# semidiscretization of the compressible Euler equations
+
+equations = CompressibleEulerEquations3D(1.4)
+
+initial_condition = initial_condition_constant
+
+boundary_conditions = Dict(
+  :all => BoundaryConditionDirichlet(initial_condition)
+)
+
+solver = DGSEM(polydeg=3, surface_flux=flux_lax_friedrichs,
+               volume_integral=VolumeIntegralWeakForm())
+
+# Mapping as described in https://arxiv.org/abs/2012.12040 but with less warping.
+function mapping(xi_, eta_, zeta_)
+  # Transform input variables between -1 and 1 onto [0,3]
+  xi = 1.5 * xi_ + 1.5
+  eta = 1.5 * eta_ + 1.5
+  zeta = 1.5 * zeta_ + 1.5
+
+  y = eta + 1/6 * (cos(1.5 * pi * (2 * xi - 3)/3) *
+                   cos(0.5 * pi * (2 * eta - 3)/3) *
+                   cos(0.5 * pi * (2 * zeta - 3)/3))
+
+  x = xi + 1/6 * (cos(0.5 * pi * (2 * xi - 3)/3) *
+                  cos(2 * pi * (2 * y - 3)/3) *
+                  cos(0.5 * pi * (2 * zeta - 3)/3))
+
+  z = zeta + 1/6 * (cos(0.5 * pi * (2 * x - 3)/3) *
+                    cos(pi * (2 * y - 3)/3) *
+                    cos(0.5 * pi * (2 * zeta - 3)/3))
+
+  return SVector(x, y, z)
+end
+
+# Unstructured mesh with 68 cells of the cube domain [-1, 1]^3
+mesh_file = joinpath(@__DIR__, "cube_unstructured_1.inp")
+isfile(mesh_file) || download("https://gist.githubusercontent.com/efaulhaber/d45c8ac1e248618885fa7cc31a50ab40/raw/37fba24890ab37cfa49c39eae98b44faf4502882/cube_unstructured_1.inp",
+                              mesh_file)
+
+mesh = P4estMesh{3}(mesh_file, polydeg=3,
+                    mapping=mapping,
+                    initial_refinement_level=1)
+
+# TODO P4EST FSP
+# # Refine bottom left quadrant of each tree to level 2
+# function refine_fn(p8est, which_tree, quadrant)
+#   if quadrant.x == 0 && quadrant.y == 0 && quadrant.z == 0 && quadrant.level < 2 && convert(Int, which_tree) == 0
+#     # return true (refine)
+#     return Cint(1)
+#   else
+#     # return false (don't refine)
+#     return Cint(0)
+#   end
+# end
+
+# Refine recursively until each bottom left quadrant of a tree has level 2
+# The mesh will be rebalanced before the simulation starts
+# refine_fn_c = @cfunction(refine_fn, Cint, (Ptr{Trixi.p8est_t}, Ptr{Trixi.p4est_topidx_t}, Ptr{Trixi.p8est_quadrant_t}))
+# Trixi.p8est_refine(mesh.p4est, true, refine_fn_c, C_NULL)
```

Comment:
Yes, FSP doesn't work yet on non-conforming meshes.

Reply:
Just out of curiosity: Have you confirmed that FSP works on the 3D `TreeMesh` with non-conforming elements? I just want to avoid that there is a fundamental issue that is related to how we do mortars that has nothing to do with your P4estMesh implementation.

Reply:
I haven't, but I think it should work perfectly fine on a rectangular mesh. The problem here is that I need to compute the contravariant vectors and inverse Jacobian on the trees and interpolate them onto the smaller elements (Kopriva et al., 2019).
I already implemented this, and FSP on non-conforming curved meshes works as expected. I broke 2D with this though, so I need a few days to prepare a new PR.

Reply:
OK